### PR TITLE
Add batch payment

### LINKED ISSUE #3713: Expenditures: As a user I can upload a data file
[Provided by Adrian via Contractors project management platform]

## Description
User has to be able to upload a .csv file and see the result of parsing process.

## Acceptance Criteria

- an option to upload .csv file (max. 400 records)
- an option to display batch owner
- an option to display uploading process icon
- an option to display clear batch button
- an option to display view all data button
- an option to display recipients counter
- an option to display batch value and multiple currencies
- an option to display "Draft" stage when uploading data
- an option to display "Locked' stage when uploading is complete

## Design

- [Figma Link - Download and upload Batch](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319781)
- [Figma Link - Uploading state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319521)
- [Figma Link - Uploaded state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319256)
- [Figma Link - Uploaded with multiple tokens](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3807%3A290638)

reject)

reject)

reject)

reject)

### PR DESCRIPTION
**Pull request by agnieszkajarosikloj**
## Description

This PR adds batch payment type. When user selects "Batch" payment type, new form section is displayed. This new section contains: 
- dropdown to upload .csv file,
- uploading process icon,
- clear batch button (when file is uploaded),
- view all data button (modal should be visible here, but it will be added with next PR),
- recipients counter,
- batch value and multiple currencies.

Screenshot 2022-08-08 at 09 44 23

[Figma Link - Download and upload Batch](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319781)
[Figma Link - Uploading state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319521)
[Figma Link - Uploaded state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319256)
[Figma Link - Uploaded with multiple tokens](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3807%3A290638)

## TODO

- form validation - will be added in next PR
- modal with data details - also in next PR

Resolves #3713

### CODE REVIEW

**Review by Julianahlk — COMMENTED** [review]
Looking good! A few questions regarding the design. I see a few discrepancies from what you have and what's on Figma: 
Screen Shot 2022-08-09 at 5 17 27 PM

1. Expenditure type - Should this text be changed to Payment type? 
2. Team - Should team be switched to From? 
3. Balance - This is not in the Figma design. 
4. Owner - Should the initial state for the owner be empty? 
Screen Shot 2022-08-09 at 5 07 10 PM

**Comment by agnieszkajarosikloj** [comment]
> Looking good! A few questions regarding the design. I see a few discrepancies from what you have and what's on Figma: Screen Shot 2022-08-09 at 5 17 27 PM
> 
> 1. Expenditure type - Should this text be changed to Payment type?
> 2. Team - Should team be switched to From?
> 3. Balance - This is not in the Figma design.
> 4. Owner - Should the initial state for the owner be empty?
> 
> Screen Shot 2022-08-09 at 5 07 10 PM


@arrenv if you could confirm, we agreed that the ExpenditureSettings section should be the same as for the other payment types

**Comment by arrenv** [comment]
> * Expenditure type - Should this text be changed to Payment type?

Yes, we eventually want the naming to be updated across the whole feature from "Expenditure" to "Payment". So, you can just replace it.

> * Team - Should team be switched to From?

No, leave it as "Team".

> * Balance - This is not in the Figma design.

This can stay there.

> * Owner - Should the initial state for the owner be empty?

No, leave it as it is, i.e. auto added based on who is creating it.

**Review by willm30 — COMMENTED** [review]
@agnieszkajarosikloj With the following CSV data I'm not getting a value for recipients or value: 

Recipient | Token | Amount
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1000
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1001
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1002
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1003
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1004
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1005
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1006
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1007
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | xDai | 1008

**Comment by willm30** [comment]
@agnieszkajarosikloj Recipient is showing up for me now but value is not. Using the same CSV.

**Comment by agnieszkajarosikloj** [comment]
> @agnieszkajarosikloj Recipient is showing up for me now but value is not. Using the same CSV.

@willm30 token in CSV file should be added by the token address/id. 
I added displaying 'Recipients' and 'Value' with message even when there are no valid data. Thanks for pointing it.

**Comment by Julianahlk** [comment]
I can't get recipients counter, batch value and multiple currencies to be shown. 

Screen Shot 2022-08-19 at 5 20 09 PM

**Comment by willm30** [comment]
@agnieszkajarosikloj I've updated my CSV to the below, using the token address in the token column (the address is for Tether):
```
Recipient | Token | Amount
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1000
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1001
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1002
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1003
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1004
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1005
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1006
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1007
0xCd96D435128415F3265c32f9eaeFaCc6Be53b7D7 | 0xdAC17F958D2ee523a2206206994597C13D831ec7 | 1008
```

I'm only seeing this view:

batch-csv

What CSV are you using in your screenshot in the issue description?

**Comment by agnieszkajarosikloj** [comment]
@willm30 I've merged PR with 'Download template' button. Could you try to download the template and then try to upload the CSV file from the example with your data? Please let me know if that works.
And just to be sure, recipient id's and tokens are valid? Because I am filtering the data: 
```
const { data: colonyMembers } = useMembersSubscription({
    variables: { colonyAddress: colony.colonyAddress || '' },
  });
const { tokens: colonyTokens } = colony || {};
    
const validatedData = data.map(({ recipient, amount, token }) => {
      const correctRecipient = colonyMembers?.subscribedUsers.find(
        (user) => user.id === recipient,
      );
      const correctToken = colonyTokens?.find(
        (tokenItem) => tokenItem.id === token,
      );
      return {
        recipient: correctRecipient ? recipient : undefined,
        amount,
        token: correctToken || undefined,
      };
    });
```

**Comment by willm30** [comment]
@agnieszkajarosikloj Got it working now. I was going wrong with the tokens (maybe a more explicit error message here, something like "token address not found in colony..." would be helpful).

This csv produces a value of `3,000` and `3` recipients. Should it be 1 recipient given that it's the same address three times? 

```csv
Recipient | Token | Amount
0xb77D57F4959eAfA0339424b83FcFaf9c15407461 | 0xC9Ef69D2bE69C3BC9D7444982597fAF1A90697d1 | 1000
0xb77D57F4959eAfA0339424b83FcFaf9c15407461 | 0xC9Ef69D2bE69C3BC9D7444982597fAF1A90697d1 | 1000
0xb77D57F4959eAfA0339424b83FcFaf9c15407461 | 0xC9Ef69D2bE69C3BC9D7444982597fAF1A90697d1 | 1000
```

**Comment by agnieszkajarosikloj** [comment]
@willm30  I'm glad it works now :)  On [this](https://github.com/JoinColony/colonyDapp/pull/3751) branch, after opening the modal, a detailed list is displayed, where you can check if some data was invalid (e.g. token was not found).
@arrenv could you please confirm if i have to check if it's the same address?

**Review by chinins — COMMENTED** [review]
The code looks fine but I can't make the csv template work (and I've tried many different options inlucding download template). I think that it would be great to include a working file for testing where we can just replace the values. Could you please add it to github either in PR description or in a comment. 

@arrenv, as a side note, I spent a very long time to try and make it work but couldn't. How are normal users supposed to use it? I would be infuriated if I was just a user and would abandon the task.

**Comment by arrenv** [comment]
> @arrenv, as a side note, I spent a very long time to try and make it work but couldn't. How are normal users supposed to use it? I would be infuriated if I was just a user and would abandon the task.

I am about to test it, do you mean the download template is not complete? The one I provided included example data in it. I think it can be improved from I provided though, this one just looks to be a blank csv.

**Comment by chinins** [comment]
> I am about to test it, do you mean the download template is not complete? The one I provided included example data in it. I think it can be improved from I provided though, this one just looks to be a blank csv.

There is no example data in the download template, unless I went mad :) So I tried to add data and submit and never managed to. Maybe I'm doing something really wrong but I don't think that's a great UX.
Screenshot 2022-09-02 at 22 48 58

**Comment by arrenv** [comment]
> There is no example data in the download template, unless I went mad :) So I tried to add data and submit and never managed to. Maybe I'm doing something really wrong but I don't think that's a great UX. 

Yeah, that is not the csv example I provided, I will provide it again.

**Comment by willm30** [comment]
@chinins I [got it working](https://github.com/JoinColony/colonyDapp/pull/3728#issuecomment-1223737449) by ensuring the tokens are an address and exist in the colony.

@arrenv  If there are three entries with the same recipient, should this show up as 1 or 3 recipients? I would have thought just 1 since it's the same address.

**Comment by arrenv** [comment]
> @chinins I [got it working](https://github.com/JoinColony/colonyDapp/pull/3728#issuecomment-1223737449) by ensuring the tokens are an address and exist in the colony.
> 
> @arrenv If there are three entries with the same recipient, should this show up as 1 or 3 recipients? I would have thought just 1 since it's the same address.

I think this comes down to more the label of the row. Ignoring that, either option seems reasonable, although, I would lean more towards showing 3, given that is what was added in the csv file.

However, perhaps the `Recipients` label should be improved to be more accurate, e.g. `Imported payments`.

**Review by willm30 — COMMENTED** [review]
@agnieszkajarosikloj Working well for me. I do agree though that the "Recipients" heading should be changed to "Imported payments" (and in the template as well).

**Review by willm30 — COMMENTED** [review]
@agnieszkajarosikloj The template still includes the heading "Recipient". I also expect you'll need to update `useCalculateBatchPayment` with the new heading.

**Comment by agnieszkajarosikloj** [comment]
@willm30 I thought that only label should be changed to "Imported payments". If we also change the header in the template, the user may not know that we need user id here.

**Review by willm30 — CHANGES_REQUESTED** [review]
@agnieszkajarosikloj I can see what you mean. Let's leave as is. 

With respect to the error state, when you upload the provided template as is, in the "Recipients" column the preview says that "the address format is wrong". I don't think that is correct. Should it not be something like "Address not found in colony"?

When a token is not found, [the row should be red](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=5937%3A331943).

I also think the  `Not all rows were imported, please review.` message should read `Not all rows were imported correctly, please review.`, to avoid confusion. Otherwise it implies that we're expecting some set amount of rows to be imported, which we aren't. 

And when that message is showing, I would still [expect to see](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=5941%3A332960) some calculation in "Imported Payments" and "Value". Currently, when uploading the template unchanged, it says "No payments" and "No tokens", even though there is data in the csv (albeit erroneous data).

File: src/modules/dashboard/components/ExpenditurePage/Batch/DownloadTemplate/constants.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable max-len */
+export const CSV = [
```
Comment (arrenv) [inline src/modules/dashboard/components/ExpenditurePage/Batch/DownloadTemplate/constants.ts:2]: Please change to updated provided CSV.

**Review by arrenv — CHANGES_REQUESTED** [review]
Thank you @agnieszkajarosikloj 

1. Could we replace the download template with this one:
[expenditures_batch.csv](https://github.com/JoinColony/colonyDapp/files/9584312/expenditures_batch.csv)


2. It does not seem to be supporting contract addresses. The following two should both say 5 imported payments and 360,000 total ETH payments:

32487698-f74533f6ab18cb0002fe91171f52b052 (1)

32488174-1feb8609c9a7280f19b4474da99afb3a



4. I have a request. Could we please add in this tooltip to help - https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=3746%3A319781

Screenshot 2022-09-16 145638


5. Could we please have the weight of this font the same as the blue font.

32487547-2ad90ec0e91219757c0e3eb8f8f5cf82


6. There is no validation on the number of rows imported:

32488058-e3c67b55e8c671ea6959936932bc12e3

**Comment by agnieszkajarosikloj** [comment]
Screenshot 2022-09-19 at 09 36 04

@arrenv validation has been added in other PR : https://github.com/JoinColony/colonyDapp/pull/3757

**Review by willm30 — COMMENTED** [review]
@agnieszkajarosikloj At the moment when uploading a csv of more than 400, not seeing an error.

**Comment by agnieszkajarosikloj** [comment]
@willm30 form validation has been added in other PR: https://github.com/JoinColony/colonyDapp/pull/3757. 
In this PR error message is displayed only if user address or token address aren't correct.
(I've changed the recipient ID validation, it doesn't have to be a member of the colony now.)

**Review by willm30 — APPROVED** [review]
@agnieszkajarosikloj Ok thanks. Good to go from me.

File: src/modules/dashboard/components/ExpenditurePage/Batch/Batch.tsx [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    id: 'dashboard.ExpenditurePage.Batch.invalidRows',
+    defaultMessage: `Not all rows were imported correctly, please review. {button}`,
+  },
+  viewAll: {
+    id: `dashboard.ExpenditurePage.Batch.viewAll`,
+    defaultMessage: 'View all',
+  },
+  tooltipText: {
+    id: `dashboard.ExpenditurePage.Batch.tooltipText`,
+    defaultMessage: `Download and use the .csv import template. Be sure to remove the sample data and view the details are correct once they have been imported.`,
```
Comment (arrenv) [inline src/modules/dashboard/components/ExpenditurePage/Batch/Batch.tsx:63]: Sorry, can we please change this to "Download and use the .csv import template. Be sure to remove the sample data and confirm the details are correct once imported."

**Review by arrenv — APPROVED** [review]
Looking and working great, just a small point. Thanks @agnieszkajarosikloj

**Pull request merged**